components/search_results/search_limits_banner.test.tsx

```diff
@@ -0,0 +1,288 @@
+    prod_1: {
+        id: 'prod_1',
+        sku: CloudProducts.STARTER,
+        price_per_seat: 0,
+        name: 'Cloud Starter',
+    },
+    prod_2: {
+        id: 'prod_2',
+        sku: CloudProducts.PROFESSIONAL,
+        price_per_seat: 10,
+        name: 'Cloud Professional',
+    },
+    prod_3: {
+        id: 'prod_3',
+        sku: CloudProducts.ENTERPRISE,
+        price_per_seat: 30,
+        name: 'Cloud Enterprise',
+    },
+};
+
+describe('components/select_results/SearchLimitsBanner', () => {
+    test('should NOT show banner for non cloud when doing messages search', () => {
+        const state = {
+            entities: {
+                general: {
+                    license: {
+                        IsLicensed: 'true',
+                        Cloud: 'false', // not cloud
+                    },
+                },
+                users: {
+                    currentUserId: 'uid',
+                    profiles: {
+                        uid: {},
+                    },
+                },
+                cloud: {
+                    subscription: null,
+                    products: null,
+                    limits: {
+                        limits: {},
+                        limitsLoaded: false,
+                    },
+                },
+                usage,
+            },
+        };
+        const store = mockStore(state);
+        const wrapper = mountWithIntl(<Provider store={store}><SearchLimitsBanner searchType='messages'/></Provider>);
+        expect(wrapper.find('#messages_search_limits_banner').exists()).toEqual(false);
+    });
+
+    test('should NOT show banner for non cloud when doing files search', () => {
+        const state = {
+            entities: {
+                general: {
+                    license: {
+                        IsLicensed: 'true',
+                        Cloud: 'false', // not cloud
+                    },
+                },
+                users: {
+                    currentUserId: 'uid',
+                    profiles: {
+                        uid: {},
+                    },
+                },
+                cloud: {
+                    subscription: null,
+                    products: null,
+                    limits: {
+                        limits: {},
+                        limitsLoaded: false,
+                    },
+                },
+                usage,
+            },
+        };
+        const store = mockStore(state);
+        const wrapper = mountWithIntl(<Provider store={store}><SearchLimitsBanner searchType='files'/></Provider>);
+        expect(wrapper.find('#files_search_limits_banner').exists()).toEqual(false);
+    });
+
+    test('should NOT show banner for non cloud when doing cloud messages search above the limit but NOT in starter product', () => {
+        const aboveMessagesLimitUsage = {
+            ...usage,
+        };
+
+        aboveMessagesLimitUsage.messages.history = 15000; // above limit of 10k
```

Comment:
This still mutates the `usage.messages`. i.e.
```js
const usage = {
    messages: {
        history: 10000,
    },
};
console.log(usage)
const newUsage = {
    ...usage,
}
newUsage.messages.history = 15000;
console.log(newUsage)
console.log(usage)
```

outputs

```
node rw/mattermost-webapp/tmp.js
{ messages: { history: 10000 } }
{ messages: { history: 15000 } }
{ messages: { history: 15000 } }
```

This happens because the spread operator only does a shallow copy. It doesn't end up causing any problems in the tests here, but its not best practice and if test run order wasn't defined and different tests needed different values and didn't reset them all, you could get mysterious test failures.

To resolve, you could either use `makeEmptyUsage` in each test and then mutate because the value is scoped to that single test, or you can more deeply copy, i.e.

```js
        const aboveMessagesLimitUsage = {
            ...usage,
            messages: {
              ...usage.messages,
              history: 15000
            }
        };

        aboveMessagesLimitUsage.messages.history = 15000;
```